### PR TITLE
Project 1: Autograd

### LINKED ISSUE #16: Project 1 Proposal: Auto-differentiation
**What will you do?**
Implement auto-differentiation for the main function. We will extend the language to take in inputs for the main function and return values. We will then output the derivatives of the outputs wrt to the inputs when a special interpreter is invoked (which supplies input values).

**How will you do it?**
There's a couple different ways. We will start by implementing forward mode auto-differentiation (with dual numbers), then reverse-mode (by constructing a graph), and finally source to source translation (e.g. Julia's Zygote.jl) if we have time.

**How will you empirically measure success?**
We will write a short script to optimize a function (run it with inputs, take a step of gradient descent, continue), and test it on arbitrary functions. In addition, we will evaluate the standard properties of forward-mode vs reverse-mode differentiation to make sure that they need the correct number of function evaluations.

With:
@q-hwang

### PR DESCRIPTION
**Pull request by Chillee**
Fixes https://github.com/sampsyo/cs6120/issues/16

With @q-hwang

### CODE REVIEW

File: content/blog/2019-09-25-autograd.md [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+the auto-differentiation must be done outside of the Bril language itself.
+One thing that would be interesting future work would be to integrate the
+function call work of one the other groups with Bril syntax so that we could
+use auto-differentiation from within Bril.
+
+Another issue that we had was in separating out the different
+auto-differentiation systems. While in other languages, AD systems can often
+be implemented through a combination of custom objects + operator
+overloading, Bril doesn't have either one of these concepts. As such, we
+simply shoved everything inside of the interpreter.
```
Comment (sampsyo) [inline content/blog/2019-09-25-autograd.md]: I'd argue that doing interpretation in the interpreter—even when it's a non-standard interpretation—is actually _more_ elegant than doing it by wonton operator overloading. :smiley:

File: content/blog/2019-09-25-autograd.md [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+if (x>0) {
+    y = x;
+} else {
+    y = 0-x;
+}
+```
+![](https://i.imgur.com/hZKqMEu.jpg)
+We see that unlike when our function was nice and convex, we constantly oscillate in this setting.
+
+We also tested it on functions with multiple input variables and output variables. Unluckily, those a bit more difficult to visualize. However, we can report some results :)
```
Comment (sampsyo) [inline content/blog/2019-09-25-autograd.md]:
```suggestion
We also tested it on functions with multiple input variables and output variables. Unluckily, those a bit more difficult to visualize. However, we can report some results. :)
```

File: content/blog/2019-09-25-autograd.md [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+
+For example, optimizing the function
```
Comment (sampsyo) [inline content/blog/2019-09-25-autograd.md]:
```suggestion
For example, optimizing the function:
```

**Review by sampsyo — CHANGES_REQUESTED** [review]
Nice work! This was a fun read. I have to admit that I had some trouble understanding the core of the discussion of AD algorithms—there's no need to go crazy rewriting that stuff to make it more intelligible, but I did leave some comments about where exactly I was confused.

I have two clerical suggestions:

- There's lots of math in this post—consider using our LaTeX setup to render that instead of math-like code snippets.
- Please include the images in this repository—it seems like something is going wrong with the Imgur hosting approach.

File: content/blog/2019-09-25-autograd.md [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+differentiation, we replace all numbers in our program with dual numbers.
+
+Then, the derivative we want of a given number wrt to the input is simply
+given by its epsilon term. So, in the addition example, the derivative of the
+addition of two numbers is equal to the sum of their derivatives. For
+multiplication, its equal to `(a*b + c*d)`.
+
+So, what is this epsilon term? One (not entirely inaccurate) way of
+interpreting it is as an infinitesimal. Squaring an infinitesimal makes it
+disappear, but otherwise, it's somewhat similar to numerical differentiation.
```
[the remark this answers is not in the document]
Comment (Chillee) [reply content/blog/2019-09-25-autograd.md]: Well... it is a common thing to do in places like numerical analysis when analyzing floating point error.

I'm not sure this is entirely the best interpretation of dual numbers. I know there's another interpretation that's connected with matrices and taylor expansion.

However, A. I don't understand dual numbers well enough to grasp those intuitively, B. it'd be significantly harder to explain.

File: content/blog/2019-09-25-autograd.md [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+
+We implemented forward-mode in this fashion. The primary work was done in the
+`brili` interpreter, where we simply interpreted every integer as a floating
+point number, and then augmented it to be a dual number. Then, we simply
+augmented every single operation to either operate on dual numbers (addition,
+multiplication, etc.) or kill the derivatives (conditionals).
```
[the remark this answers is not in the document]
Comment (Chillee) [reply content/blog/2019-09-25-autograd.md]: Added an explanation.

File: content/blog/2019-09-25-autograd.md [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+
+`dz/dy dy/dx = f'(y)g'(x) = f'(g(x))g'(x)`.
+
+Now, back to automatic differentiation.
+
+Given `dy/dx = dy/dw_2 dw_2/dw_1 dw_1/x`, there are two ways we can decompose
+this expression into functions. We could compute `dy/dx = (dy/dw_2)
+(dw_2/dw_1 dw_1/x)` or we could compute `dy/dx = (dy/dw_2 dw_2/dw_1)
+(dw_1/x)`. The first one is forward-mode automatic differentiation, the
+second one is reverse-mode.
```
[the remark this answers is not in the document]
Comment (Chillee) [reply content/blog/2019-09-25-autograd.md]: I think I might move this paragraph around. I think the one way to think about the difference is that we can only compute either prefixes or suffixes of this long chain of derivatives (where a long chain corresponds to the derivative of one output wrt to an input). When there's only one input and one output, this performs the same amount of computation. However, when you have multiple input and outputs, these chains have the potential to overlap. Depending on whether your overall derivative graph expands from left to right (ie: you have more outputs than inputs), or it contracts (ie: you have more inputs than outputs), you have the potential to save on computation.

However, this is a bit hard to explain without talking about the computational properties.

File: content/blog/2019-09-25-autograd.md (hunk 1 above)
Comment (Chillee) [reply content/blog/2019-09-25-autograd.md]: It's more elegant perhaps - but it's also nice to be able to separate autodifferentiation from your core language :)

Comment (Chillee) [reply content/blog/2019-09-25-autograd.md]: I actually still think operator overloading is nicer - it fits nicer with my model of what's going on. You have your regular data types in your language, and then you have your "derivative enabled" data types.

File: content/blog/2019-09-25-autograd.md [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+we're flipping the forward-mode differentiation algorithm on its head.
+
+This allows us to compute an output variable wrt arbitrarily many input
+variables in merely a single pass through the function. That is, we need
+`O(M)` evaluations of a function `f: R^N -> R^M` (as opposed to `O(N)` for
+forward-mode).
+
+Implementing this is a bit trickier than implementing forward-mode. We can no longer tie our gradient computation to our actual computation, we must construct a graph that allows us to *replay* our derivatives after all of our computation is finished.
+
+We implemented this by constructing a `RevADNode` for every computation we perform. This `RevADNode` is pushed onto the children of the inputs to this computation. In this way, our entire computation is saved as a graph in memory, while no gradients are computed. At each point, we simply need to store the value of the expression, as well as its portion of the derivative.
```
[the remark this answers is not in the document]
Comment (Chillee) [reply content/blog/2019-09-25-autograd.md]: Yeah that's a good way to make it clear.

File: content/blog/2019-09-25-autograd.md [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+
+## Reverse-Mode Automatic Differentiation
+
+Reverse-mode automatic differentiation is less intuitive than forward-mode.
+So why do we need reverse-mode in the first place? So forward-mode allows us
+to compute arbitrary derivatives, but notice that in order to get how the
+output varied with the 2 input variables, we needed to apply the
+auto-differentiation algorithm twice. In particular, forward-mode
+auto-differentiation requires `O(N)` evaluations of a function `f: R^N
+-> R^M` in order to calculate all the derivatives. This kind of function
```
[the remark this answers is not in the document]
Comment (Chillee) [reply content/blog/2019-09-25-autograd.md]: Yeah, that's right.

The auto-differentiation community typically phrases it in terms of how often the function needs to get evaluated, I believe. 

I'll clarify.

[the remark this answers is not in the document]
Comment (Chillee) [reply content/blog/2019-09-25-autograd.md]: I think it's more clear if I write it as "requires computation equivalent to `O(N)` evaluations of the function`.

File: content/blog/2019-09-25-autograd.md [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+shows up a lot in machine learning, where we're often optimizing for a single
+loss. In that case, we often want `f:R^(millions of parameters) -> R^1`.
+Needing to evaluate the function millions of times to just perform a single
+step of gradient descent is too difficult.
+
+To resolve this, we use reverse-mode automatic differentiation. At a high
+level, if forward-mode is asking the question "What input variables affect
+the current output variable?", then reverse-mode is asking the question "What
+output variables can our current input variable affect?". In some sense,
```
[the remark this answers is not in the document]
Comment (Chillee) [reply content/blog/2019-09-25-autograd.md]: Yeah my language is a bit imprecise here. Will modify.

File: content/blog/2019-09-25-autograd.md [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+#### Q: What we wrote after applying the chain rule looks awfully close to actual code. Can we generate code that generates a function that generates derivatives?
+
+A: Yes! This is a cool method of doing automatic differentiation, recently
+popularized by [Julia](https://arxiv.org/abs/1810.07951). In practice, this
+is a lot more difficult than my example may make it seem. Handling
+expressions is relatively straight-forward, but doing a source to source
+transformation that preserves control flow, state, and other such things is
+substantially more difficult. The primary pitch for doing things this way is
+that we can then take advantage of all the compiler optimizations we usually
+apply on our code.
```
[the remark this answers is not in the document]
Comment (Chillee) [reply content/blog/2019-09-25-autograd.md]: Hmm, actually, I don't think this point is particularly relevant in the forward-mode differentiation section. From my understanding, this is actually only relevant to the reverse-mode section.

**Comment by Chillee** [comment]
Responded to comments and added some more clarification when needed.

I'm still not particularly happy with some of the sections - particularly our intuitive explanations of reverse-mode. I may circle back to this later.

File: content/blog/2019-09-25-autograd.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (sampsyo) [reply content/blog/2019-09-25-autograd.md]: That makes sense to me!

File: content/blog/2019-09-25-autograd.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (sampsyo) [reply content/blog/2019-09-25-autograd.md]: Yeah, sounds good.

File: content/blog/2019-09-25-autograd.md (hunk 1 above)
Comment (sampsyo) [reply content/blog/2019-09-25-autograd.md]: However, fortunately/unfortunately, most languages don't let you overload `for`…

**Comment by sampsyo** [comment]
Looks good! I think this is a perfectly fine explanation for this context. :+1: It's a surprisingly strange topic with its own internal vocabulary, so it's deceptively difficult to explain.

**Pull request merged**